operator/pkg/metrics/server.go

```diff
@@ -0,0 +1,211 @@
+
+type Metrics struct {
+	gauges map[string]prometheus.Gauge
+}
+
+var metrics Metrics
+
+var instance *serverManager
+var smOnce sync.Once
+
+type serverManager struct {
+	server        *http.Server
+	ticker        *clock.Ticker
+	ofClient      *openfeature.Client
+	exposeMetrics bool
+}
+
+func StartServerManager(ctx context.Context, exposeMetrics bool) {
+	smOnce.Do(func() {
+		metrics.gauges = make(map[string]prometheus.Gauge)
+		instance = &serverManager{
+			ticker:        clock.New().Ticker(10 * time.Second),
+			ofClient:      openfeature.NewClient("klt"),
+			exposeMetrics: exposeMetrics,
+		}
+		instance.Start(ctx)
+	})
+}
+
+func (m *serverManager) Start(ctx context.Context) {
+	go func() {
+		for {
+			select {
+			case <-ctx.Done():
+				if err := m.ShutDownServer(); err != nil {
+					klog.Errorf("Error during server shutdown: %v", err)
+				}
+				return
+			case <-m.ticker.C:
+				if err := m.setup(); err != nil {
+					klog.Errorf("Error during server setup: %v", err)
+				}
+			}
+		}
+	}()
+}
+
+func (m *serverManager) ShutDownServer() error {
+	defer func() {
+		m.server = nil
+	}()
+	if m.server != nil {
+		ctx, cancel := context.WithTimeout(context.Background(), 1*time.Second)
+		defer cancel()
+		return m.server.Shutdown(ctx)
+	}
+	return nil
+}
+
+func (m *serverManager) setup() error {
+	maxRetries := 3
+	var serverEnabled bool
+	var err error
+
+	klog.Infof("Checking configuration of keptn-metrics server")
+
+	for i := 0; i < maxRetries; i++ {
+		serverEnabled, err = m.ofClient.BooleanValue(context.TODO(), "keptn.gms.expose", m.exposeMetrics, openfeature.EvaluationContext{})
+		if err == nil {
+			break
+		}
+
+		if strings.Contains(err.Error(), string(openfeature.ProviderNotReadyCode)) {
+			<-time.After(2 * time.Second)
+			continue
+		}
+		break
+	}
+
+	klog.Infof("Keptn Metrics server enabled: %v", serverEnabled)
+
+	if serverEnabled && m.server == nil {
+		klog.Infof("serving metrics at localhost:9999/metrics")
+
+		router := mux.NewRouter()
+		router.Path("/metrics").Handler(promhttp.Handler())
+		router.Path("/api/v1/metrics/{namespace}/{metric}").HandlerFunc(returnMetric)
+
+		m.server = &http.Server{
+			Addr:    ":9999",
+			Handler: router,
+		}
+
+		m.recordMetrics()
+
+		go func() {
+			err := m.server.ListenAndServe()
+			if err != nil {
+				klog.Errorf("could not start keptn-metrics server: %w", err)
+			}
+		}()
+
+	} else if !serverEnabled && m.server != nil {
+		if err := m.ShutDownServer(); err != nil {
+			return fmt.Errorf("could not shut down keptn-metrics server: %w", err)
+		}
+	}
+	return nil
+}
+
+func returnMetric(w http.ResponseWriter, r *http.Request) {
+	vars := mux.Vars(r)
+	namespace := vars["namespace"]
+	metric := vars["metric"]
+
+	scheme := runtime.NewScheme()
+	if err := metricsv1alpha1.AddToScheme(scheme); err != nil {
+		fmt.Println("failed to add keptn-metrics to scheme: " + err.Error())
+	}
+	cl, err := ctrlclient.New(config.GetConfigOrDie(), ctrlclient.Options{Scheme: scheme})
```

Comment:
Let's make the k8s client a property of the ServerManager that can be passed to its start function, and use that one instead of initialising it within this function